### PR TITLE
Enable ILogger by default

### PR DESCRIPTION
**Pull request by cijothomas**
Fix Issue #854 
1. Take dependency of Microsoft.Extensions.Logging.ApplicationInsights for NETSTANDARD2.0
2. Enables ApplicationInsightsILogger Provider by default for Warning or above logs.
3. Removes the default DebugLogger - its irrelevant now, as Ilogger logs are now automatically captured.
4. Add FunctionalTests to validate ILogger integration.

- [ ] I ran Unit Tests locally.

For significant contributions please make sure you have completed the following items:

- [ ] Changes in public surface reviewed
- [ ] Design discussion issue #
- [ ] CHANGELOG.md updated with one line description of the fix, and a link to the original issue.
- [ ] The PR will trigger build, unit tests, and functional tests automatically. If your PR was submitted from fork - mention one of committers to initiate the build for you.
	  If you want to to re-run the build/tests, the easiest way is to simply Close and Re-Open this same PR. (Just click 'close pull request' followed by 'open pull request' buttons at the bottom of the PR)

- Please follow [these] (https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/develop/Readme.md) instructions to build and test locally.

### CODE REVIEW

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs [hunk 1]
```diff
@@ -67,26 +67,30 @@ public static ILoggerFactory AddApplicationInsights(this ILoggerFactory factory,
             Func<string, LogLevel, bool> filter,
             Action loggerAddedCallback)
         {
-            var client = serviceProvider.GetService<TelemetryClient>();
-            var debugLoggerControl = serviceProvider.GetService<ApplicationInsightsLoggerEvents>();
-            var options = serviceProvider.GetService<IOptions<ApplicationInsightsLoggerOptions>>();
-
-            if (options == null)
+            var loggerMarker = serviceProvider.GetService<LoggerMarker>();
```
Comment (cijothomas) [inline src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: ConfigureServices() is guaranteed to be called before Configure() where this method is invoked, ensuring that LoggerMarker will be present by the code this code is executed.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj [hunk 2]
```diff
@@ -95,10 +95,11 @@
...
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
-  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.DiagnosticAdapter" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.0.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.DiagnosticAdapter" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Logging.ApplicationInsights" Version="2.9.1" />
```
Comment (cijothomas) [inline src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: This also has Microsoft.ApplicationInsights dependency. So need to see how to avoid conflicts between this and the direct dependency for Microsoft.ApplicationInsights

**Comment by cijothomas** [comment]
@SergeyKanzhelev @pakrym @RamjotSingh Please review when you get a chance! thanks in advance.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs (hunk 1 above)
Comment (RamjotSingh) [inline src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: Instead of doing this, maybe we can do IStartupFilter?

Comment (RamjotSingh) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: Essentially above in #If NetStandard2.0 we can put an additional #else and add a startup filter for starting up the old path.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs [hunk 3]
```diff
@@ -7,7 +7,10 @@
     /// <summary>
     /// <see cref="ILoggerProvider"/> implementation that creates returns instances of <see cref="ApplicationInsightsLogger"/>
     /// </summary>
+#if !NETSTANDARD2_0
```
Comment (RamjotSingh) [inline src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs]: For clarity sake, put the whole class in #IF?

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs [hunk 4]
```diff
@@ -201,12 +201,28 @@ public static IServiceCollection AddApplicationInsightsTelemetry(this IServiceCo
...
+                // to detect ApplicationInsightsLoggerProvider as already added and back-off.
+                // Without this, there will be double logging if LoggerFactory.AddApplicationInsights() was already used.
+#if NETSTANDARD2_0
+                services.AddLogging(loggingBuilder =>
+                {
+                     loggingBuilder.AddApplicationInsights();
+
+                     // By default, all logs Warning or above is captured.
+                     // AddFilter is additive
+                     loggingBuilder.AddFilter<Microsoft.Extensions.Logging.ApplicationInsights.ApplicationInsightsLoggerProvider>("Default", LogLevel.Warning);
```
Comment (RamjotSingh) [inline src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: I wonder if we are messing with the ILogger infra here. This can already be done from the config which is read as part of ILogger setup. Hiding this statement this deep in SDK might be non-intuitive.

Comment (bradygaster) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: I'd be interested in @davidfowl's opinion here. Ideally, custom middleware authors would wire up all of the services requisite for their expected behavior. If someone uses `AddApplicationInsights`, for instance, we'd presume all trace, exception - everything, by default - goes to Application Insights, and that any service wire-ups required for that goal to be "easy" or even "invisible" to the developer.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs [hunk 5]
```diff
@@ -7,7 +7,10 @@
...
+    // For NETSTANDARD2.0 We take dependency on Microsoft.Extensions.Logging.ApplicationInsights which has ApplicationInsightsProvider having the same ProviderAlias and don't want to clash with this ProviderAlias.
     [ProviderAlias("ApplicationInsights")]
+#endif
     internal class ApplicationInsightsLoggerProvider : ILoggerProvider
```
Comment (pakrym) [inline src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs]: Why do you keep this type instead of completely switching to Microsoft.Extensions.Logging.ApplicationInsights implementation?

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj [hunk 6]
```diff
@@ -86,19 +86,23 @@
...
 
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
-  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.DiagnosticAdapter" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.0.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.DiagnosticAdapter" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.1.0" />
```
Comment (pakrym) [inline src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Do you need this reference? I would be brought in by `Microsoft.Extensions.Logging.ApplicationInsights`

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs (hunk 4 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: Since users updatng to this new SDK version has not explicitly made a decision to send ILogger logs to AI, i'd like to just sent Warning or above to AI. Without this, we could end up sending too much logs to AI. (negative surprises when billing comes). Since the runtime itself logs using ILogger, we have experienced user complaints about too much AI logs. (https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues/603)

I'd open to all suggestions here. We want to balance between 'easy/auto capture of ILogger logs' and 'too much noise'.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs (hunk 1 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: @RamjotSingh Is there any scenario broken with this? Or just a suggestion?

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs (hunk 3 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs]: We cannot remove this whole class - as that'd be a breaking change for AI customers, and we dont want that. (except when we change major from to 3.XX)..

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs (hunk 1 above)
Comment (RamjotSingh) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: Probably not broken (not sure). But this code looked like it was abusing a DI construct. Hence the suggestion.

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs (hunk 4 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: By default, only "Warning" or above LogLevel logs are sent to ApplicationInsightsProvider for all categories. But this can be easily configured by user either in code or using appsettings.json.  
I added more description here: https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues/854

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs (hunk 5 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerProvider.cs]: We cannot remove this whole class - as that'd be a breaking change for AI customers, and we don't want that. (except when we change major from to 3.XX).

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 6 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: right.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs (hunk 1 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: This need revisit. By doing this, i am essentially breaking behavior already - i.e users already using logerFactory.AddApplicationInsights will see that their Filters are not respected, as this PR now favors the new Provider.

Need to do the inverse somehow - i.e for existing loggerFactory.AddApplicationInsights users, the new Provider should be disabled.

Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: We don't need to detect this actually. Since the old ILoggerFactory method is marked obsolete, we expect customers to stop using it. I will improve the Obsolete warning to be more descriptive and link to official docs.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 2 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: This is not a new issue.Same with taking dependency of Perf/Dep packages. We now need to ensure loggingis also updated with latest base sdk.

File: src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs (hunk 1 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLoggerFactoryExtensions.cs]: closing this as not longer applicable.

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs [hunk 7]
```diff
@@ -201,12 +201,38 @@ public static IServiceCollection AddApplicationInsightsTelemetry(this IServiceCo
...
+                    // }
+                    // },
+                    // The reason is as both rules will match the filter, the last one added wins.
+                    // To ensure that the default filter is in the beginning of filter rules, so that user override from Configuration will always win, 
+                    // we add code filter rule to the 0th position as below.
+
+                    loggingBuilder.Services.Configure<LoggerFilterOptions>
+                    (options => options.Rules.Insert(0,
+                        new LoggerFilterRule(
+                            "Microsoft.Extensions.Logging.ApplicationInsights.ApplicationInsightsLoggerProvider", "",
```
Comment (pakrym) [inline src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: The category is usually set to null for "all categories" case.

Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: fixed.

**Comment by cijothomas** [comment]
@SergeyKanzhelev can you review?

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs [hunk 8]
```diff
@@ -406,8 +432,8 @@ public static IServiceCollection AddApplicationInsightsTelemetryProcessor<T>(thi
 
         private static bool IsApplicationInsightsAdded(IServiceCollection services)
```
Comment (SergeyKanzhelev) [inline src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: what is this method used for? Why the change?

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj [hunk 9]
```diff
@@ -86,17 +86,20 @@
     <PackageReference Include="Microsoft.ApplicationInsights.PerfCounterCollector" Version="2.10.0-beta2" />
     <PackageReference Include="Microsoft.ApplicationInsights.WindowsServer" Version="2.10.0-beta2" />
     <PackageReference Include="Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel" Version="2.10.0-beta2" />
-    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="1.0.2" />
-    <PackageReference Include="Microsoft.Extensions.Configuration" Version="1.0.2" />
+    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="1.0.2" />
```
Comment (SergeyKanzhelev) [inline src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: will `Microsoft.AspNetCore.Hosting.Abstractions` be enough?

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj [hunk 10]
```diff
@@ -86,17 +86,20 @@
...
-    <PackageReference Include="Microsoft.Extensions.Configuration" Version="1.0.2" />
+    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="1.0.2" />           
+    <PackageReference Include="System.Text.Encodings.Web" Version="4.3.1" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' != 'netstandard2.0' ">
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="1.0.2" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="1.0.2" />
-    <PackageReference Include="System.Text.Encodings.Web" Version="4.3.1" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="1.0.2" />
```
Comment (RamjotSingh) [inline src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Same as above, can .abstractions be enough?

Comment (RamjotSingh) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Actually this should already be getting imported coz of Microsoft.Extensions.Configuration.Json

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj [hunk 11]
```diff
@@ -86,17 +86,20 @@
...
   </ItemGroup>
 
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
-  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.0.0" />
-  <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.0.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.1.0" />
+  <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.1.0" />
```
Comment (RamjotSingh) [inline src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: This should also be coming from Microsoft.Extensions.Logging.AI so you can skip it here.

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs (hunk 8 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: we use this to detect if AI services are already added to DI to prevent double injection if customer calls both AddApplicationInsightsTelemetry() and UseApplicationInsights(). We previously relied on presence of 'ApplicationInsightsDebugLogger' to detect this. With that no longer needed (since ilogger is automatically enabled), we are now relying on presense of telemetryclient to detect if AI is already added.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 9 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Likely yes, but i don't want to make un-related changes to this PR. Will revisit all package references in a separate PR.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 10 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Likely yes, but i don't want to make un-related changes to this PR. Will revisit all package references in a separate PR.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 11 above)
Comment (cijothomas) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: Likely yes, but i don't want to make un-related changes to this PR. Will revisit all package references in a separate PR.

File: src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj (hunk 10 above)
Comment (davidfowl) [reply src/Microsoft.ApplicationInsights.AspNetCore/Microsoft.ApplicationInsights.AspNetCore.csproj]: We should do that because a bunch of these dependencies look suspect

File: src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs [hunk 12]
```diff
@@ -201,12 +201,38 @@ public static IServiceCollection AddApplicationInsightsTelemetry(this IServiceCo
...
-                services.AddSingleton<ApplicationInsightsLoggerEvents>();
+                services.AddSingleton<JavaScriptSnippet>();                
 
                 services.AddOptions();
                 services.AddSingleton<IOptions<TelemetryConfiguration>, TelemetryConfigurationOptions>();
                 services.AddSingleton<IConfigureOptions<TelemetryConfiguration>, TelemetryConfigurationOptionsSetup>();
+
+                // NetStandard2.0 has a package reference to Microsoft.Extensions.Logging.ApplicationInsights, and
+                // enables ApplicationInsightsLoggerProvider by default.                
+#if NETSTANDARD2_0
```
[the remark this answers is not in the document]
Comment (pharring) [reply src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs]: NetFX will always "win" over netstandard2.0
